tests/test_intervals.py

```diff
@@ -358,3 +363,163 @@ def test_getitem_slice_errors(self):
+
+    def test_read_hapmap_gzipped(self, tmp_path):
+        hapfile = os.path.join(tmp_path, "hapmap.txt.gz")
+        with gzip.GzipFile(hapfile, "wb") as gzfile:
+            gzfile.write(b"HEADER\n")
+            gzfile.write(b"chr1 0 1\n")
+            gzfile.write(b"chr1 1 5.5\n")
+            gzfile.write(b"chr1 2 0\n")
+        rm = msprime.read_hapmap(hapfile, rate_col=2)
+        np.testing.assert_array_equal(rm.position, [0, 1, 2])
+        np.testing.assert_array_equal(rm.rate, [1e-8, 5.5e-8])
+
+    def test_read_hapmap_nonzero_map_start(self):
+        hapfile = io.StringIO(
+            """\
+            HEADER
+            chr1 1 x 0.000001
+            chr1 2 x 0.000001 x
+            chr1 3 x 0.000006 x x x"""
+        )
+        rm = msprime.read_hapmap(hapfile)
+        print(rm.position, rm.rate)
+        np.testing.assert_array_equal(rm.position, [0, 1, 2, 3])
+        assert np.allclose(rm.rate, [1e-8, 0, 5e-8])
+
+    def test_read_hapmap_bad_nonzero_map_start(self):
+        hapfile = io.StringIO(
+            """\
+            HEADER
+            chr1 0 x 0.0000005
+            chr1 1 x 0.000001 x
+            chr1 2 x 0.000006 x x x"""
+        )
+        with pytest.raises(ValueError, match="start.*must be zero"):
+            msprime.read_hapmap(hapfile)
+
+    def test_read_hapmap_string_warning(self):
+        hapfile = io.StringIO(
+            """\
+            HEADER
+            chr1 0 x 0
+            chr1 1 x 0.000001 x
+            chr10 2 x 0.000006 x x x"""
+        )
+        with warnings.catch_warnings(record=True) as w:
+            warnings.simplefilter("always")
+            rm = msprime.read_hapmap(hapfile)
+            assert len(w) == 1
+            assert "chromosome names" in str(w[0].message)
+        np.testing.assert_array_equal(rm.position, [0, 1, 2])
+        assert np.allclose(rm.rate, [1e-8, 5e-8])
+
+    def test_sequence_length(self):
+        hapfile = io.StringIO(
+            """\
+            HEADER
+            chr1 0 x 0
+            chr1 1 x 0.000001 x
+            chr1 2 x 0.000006 x x x"""
+        )
+        # test identical seq len
+        rm = msprime.read_hapmap(hapfile, sequence_length=2)
+        np.testing.assert_array_equal(rm.position, [0, 1, 2])
+        assert np.allclose(rm.rate, [1e-8, 5e-8])
+        assert rm.end_position == 2
+
+        hapfile.seek(0)
+        rm = msprime.read_hapmap(hapfile, sequence_length=10)
+        np.testing.assert_array_equal(rm.position, [0, 1, 2, 10])
+        assert np.allclose(rm.rate, [1e-8, 5e-8, 0])
+        assert rm.end_position == 2
+
+    def test_bad_sequence_length(self):
+        hapfile = io.StringIO(
+            """\
+            HEADER
+            chr1 0 x 0
+            chr1 1 x 0.000001 x
+            chr1 2 x 0.000006 x x x"""
+        )
+        with pytest.raises(ValueError, match="sequence_length"):
+            msprime.read_hapmap(hapfile, sequence_length=1.999)
```

Comment:
It would be good to add a test in which we have a fragment of a real hapmap file  (e.g., from HapMap, say 10 lines) and we verify that we get approximately the same values for the rate in the returned RateMap object when we use either the rate or the map columns.